README
	
= Introduction =

Please visit the following websites where you can find more
information:

Project Honey Pot website
http://www.projecthoneypot.org/?rf=39586

Plugin cpg-httpbl
http://code.google.com/p/cpg-dnsbl/

= Contact =

Author: Mario Oyorzabal Salgado
Blog: http://blog.tuxsoul.com
E-mail: <tuxsoul@tuxsoul.com>

= Bugs =

If you find any bugs or have any idea for a feature can be usefull to this 
plugin, you can leave it in the bugtrack:

http://code.google.com/p/cpg-dnsbl/issues/list

= License =

GNU General Public License v2
http://www.gnu.org/licenses/old-licenses/gpl-2.0.html

= Requirements =

  * Key from "Project Honey Pot"
  * Quicklink from "Project Honey Pot"
  * Coppermine Gallery 1.4 +
  * PHP4 +

= Features =

  * Use Http:BL to detect spammer's
  * Ability to redirect into a trap if the visitor is a spammer
  * Ability to insert addresses traps in the html code of
    Coppermine Gallery

= Details =

  * First download and decompresses the file plugin of this
  * Rename the directory the this plugin to "cpg-httpbl"
  * Inside the directory "cpg-httpbl" there is a file named 
    "config.httpbl.php", which is necessary to change the following:

     * Enter your "key" from Project Honey Pot
     * Enter a QuickLink or trap, to redirect the spammer's
     * Enter a word random, to anything, is only to show the Links to spammer's
       traps

  * Upload the "cpg-httpbl" to the directory of plugins in your gallery, which 
    is called "plugins"
  * Using your favorite browser, go to you gallery, enter as Administrator
  * Verify that the configuration of this gallery you have enabled "Enable
    Plugins" option
  * Go to the manager Plugins: "pluginmgr.php"
 
    Ex.
        http://www.example.com/pluginmgr.php
        http://www.example.com/cpg/pluginmgr.php

  * On the screen and should appear this Plugin "cpg-httpbl"
  * Install the plugin, by clicking the letter "i", which is located to one side
    Right
 
Lastly and can only wait to block the spammer's coming to you Website.

= Donation =

Please consider helping through donations, for more information visit the 
following link:

Donate
http://tinyurl.com/36mzoa

Plugin cpg-httpbl
http://code.google.com/p/cpg-dnsbl/


Cheers :-).


CHANGELOG
================================================================================
CHANGELOG
================================================================================

--------------------------------------------------------------------------------
2008-01-26  -  v0.1
--------------------------------------------------------------------------------
* Core version '0.1'
* First version of this plugin
* Code of project wiki-httpbl
* Add codebase.php and configuration.php to work in Coppermine Gallery
* Detect spammer's using http:bl from 'Project Honey Pot'
* Can redirect spammer to one address trap
* Use one word with address trap
* Insert three or more address trap in the html code

================================================================================


httpbl.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

// Check to run with mediawiki
if (!defined('IN_COPPERMINE')) {
	die ('You are lost ?');
}

// Define path of extension
define('HTTPBL_CWD', dirname(__FILE__));

// Include class to work dnsbl of httpbl
include(HTTPBL_CWD . '/object.httpbl.php');

?>


codebase.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

if (!defined('IN_COPPERMINE')) die('Not in Coppermine...');

define ('CPG_HTTPBL_CWD', dirname(__FILE__));

// project honey pot core
include (CPG_HTTPBL_CWD . '/httpbl.php');
// Instance object api project honey pot
$httpbl = new objectHttpBL();
// If visitor evil ?
$httpbl->httpblStart();

// Add a filter for the gallery
// Show url traps in the code
$thisplugin->add_filter('gallery_header','cpg_httpbl_traps_top');
$thisplugin->add_filter('gallery_footer','cpg_httpbl_traps_footer');

// Filters code
// Add url traps in the top
function cpg_httpbl_traps_top (&$html) {
	return cpg_httpbl_traps() . $html;
}

// Add url traps in the footer
function cpg_httpbl_traps_footer (&$html) {
	return cpg_httpbl_traps() . $html;
}

// Add url traps in the html code
function cpg_httpbl_traps () {
	$httpbl = new objectHttpBL();
	return $httpbl->httpblGetUrlTrap();
}

?>


core.httpbl.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die ('You are lost ?');
}

// main class object
class coreHttpBL 
{
	var $coreHttpBLIp;
	var $coreHttpBLError;

	var $coreHttpBLConfVersion;
	var $coreHttpBLConfDnsbl;
	var $coreHttpBLConfDays;

	function coreHttpBL () {
		// set flag error to cero
		$this->coreHttpBLError = 0;
	
		// config
		// version of this core
		$this->coreHttpBLConfVersion = '0.1';

		// spammer days since last activity
		$this->coreHttpBLConfDays = '15';

		// define query to dnsbl
		$this->coreHttpBLConfDnsbl = '{key}.{ip}.dnsbl.httpbl.org';
	}

	// return version of this core
	function coreHttpBLVersion () {
		return $this->coreHttpBLConfVersion;
	}

	// start the search to spammer's
	function coreHttpBLStart () {
		// detect ip
		$this->coreHttpBLDetectIp();

		// reverse-octet ip address
		// change data of query
		$host = $this->coreHttpBLDnsblPrepareQuery();

		// make query and get response
		$response = $this->coreHttpBLDnsblQuery($host);

		// Is spammer ?
		if($this->coreHttpBLIsSpammer($response) == 1) {
			// Get the trap
			$this->coreHttpBLShowTrap();
		}
	}

	// try to detect the remote ip
	function coreHttpBLDetectIP () {
		$ip = $_SERVER['REMOTE_ADDR'];

		// only for test
		// $ip = '127.1.1.1';

		$this->coreHttpBLIp = $ip;
	}

	// change data of query to dnsbl
	function coreHttpBLDnsblPrepareQuery () {
		// reverse-octet ip format
		$reverseIp = implode('.', array_reverse(explode('.', $this->coreHttpBLIp)));

		// prepare query to dnsbl
		$string = $this->coreHttpBLConfDnsbl;
		$string = ereg_replace('{key}', HTTPBL_KEY, $string);
		$string = ereg_replace('{ip}', $reverseIp, $string);

		return $string;
	}

	// make the query to dnsbl server
	function coreHttpBLDnsblQuery ($host) {
		$response = gethostbyname($host);

		if($response == $host)
			$this->coreHttpBLError = 1;

		return $response;
	}

	// is spammer ?
	function coreHttpBLIsSpammer ($response) {
		if($this->coreHttpBLError == 0) {
			$data = explode('.', $response);
			$spammer = 0;

			// the first octect always is 127
			if((int)$data[0] == 127) {
				// spammer days since last activity
				if((int)$data[1] <= $this->coreHttpBLConfDays) {
					// only check the type of spammer
					if((int)$data[3] > 0) {
						$spammer = 1;
					}
				}		
			}
			else
				$this->coreHttpBLError = 1;

			return $spammer;
		}

		return 0;
	}

	// redirect to the trap
	function coreHttpBLShowTrap () {
		if($this->coreHttpBLError == 0) {
			// FIXME: HTTPBL_TRAP is set correctly ?
			if(strlen(HTTPBL_TRAP) > 0) {
				header('Location: ' . HTTPBL_TRAP);
				die();
			}
		}
	}

	// return url to trap in html
	// FIXME: Need more work here :)
	function coreHttpBLGetUrlTrap () {
		$text = '<!-- begin links ' . HTTPBL_TRAP_WORD . ' -->';

		// FIXME: HTTPBL_TRAP is set correctly ?
		if(strlen(HTTPBL_TRAP) > 0) {
			$url = array (
					'<a href="{url}"><!-- {word} --></a>',
					'<a href="{url}" style="display: none;">{word}</a>',
					'<div style="display: none;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"></a>',
					'<!-- <a href="{url}">{word}</a> -->',
					'<div style="position: absolute; top: -250px; left: -250px;"><a href="{url}">{word}</a></div>',
					'<a href="{url}"><span style="display: none;">{word}</span></a>',
				);

			for($x = 0; $x <= 2; $x++) {
				$temp = $url[rand(0, 6)];
				$temp = ereg_replace('{url}', HTTPBL_TRAP, $temp);
				$temp = ereg_replace('{word}', HTTPBL_TRAP_WORD, $temp);

				$text .= $temp;
			}

			$text .= '<!-- end links '. HTTPBL_TRAP_WORD . ' -->';
		}

		return $text;
	}
}
?>


config.httpbl.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

// you can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die ('You are lost ?');
}

////////// Configuration
////////// You can edit all values ;-)

///// Key to access dnsbl service from "Project Honey Pot"
///// All Access Keys are 12-characters in length, lower case, and contain only alpha characters (no numbers)
///// You need a key ?, register here: http://www.projecthoneypot.org?rf=39586
define('HTTPBL_KEY', '------------');

////////// Url to honeypot trap for redirect malicious visitors
///// If you have install a trap or you can get this in "Project Honey Pot" -> "QuickLink"
///// This value is important to catch more spammers ;-)
define('HTTPBL_TRAP', '');

///// Is important change this value to any word, for have usefull url trap's
define('HTTPBL_TRAP_WORD', 'coppermine-gallery');

?>


object.httpbl.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

// can't execute directly this file
// you are lost ?
if (!defined('HTTPBL_CWD')) {
	die('You are lost ?');
}

// include config file
include (HTTPBL_CWD . '/config.httpbl.php');

// include core file
include (HTTPBL_CWD . '/core.httpbl.php');

// main class object
class objectHttpBL extends coreHttpBL
{
	// load configuration
	function httpbl () {

	}

	// start the search to spammer's
	function httpblStart () {
		$this->coreHttpBLStart();
	}

	// get url to trap
	function httpblGetUrlTrap () {
		return $this->coreHttpBLGetUrlTrap();
	}
}
?>


configuration.php
<?php
/*
        ******************************************************************************
        CPG-HttpBL Plugin
        ******************************************************************************
        Author: Mario Oyorzabal Salgado
        Blog:   http://blog.tuxsoul.com/
        E-mail: <tuxsoul@tuxsoul.com>
        Web:    http://code.google.com/p/cpg-dnsbl/
        Donate: http://tinyurl.com/36mzoa
        ******************************************************************************
        This program is free software; you can redistribute it and/or modify
        it under the terms of the GNU General Public License as published by
        the Free Software Foundation; either version 2 of the License, or
        (at your option) any later version.
        ******************************************************************************
*/

if (!defined('IN_COPPERMINE')) die('Not in Coppermine...');

$name='CPG-HttpBL Plugin';
$description='Get out spammers of you gallery using Project Honey Pot (http:BL).';
$author='Mario Oyorzabal Salgado';
$version='0.1';

/*
 * $extra_info is displayed with the title of a plugin that is NOT installed and
 * can be used to present extra information.  In this case I show a complex
 * example that forms a button the user can click.
 */
$extra_info = <<<EOT
    <table border="0" cellspacing="0" cellpadding="0"> 
    <tr>
        <td class="admin_menu"><a target="_blank" href="http://code.google.com/p/cpg-dnsbl/" title="Home">Website</a></td>
        <td class="admin_menu"><a target="_blank" href="http://tinyurl.com/36mzoa" title="Donate">Donate</a></td>
        <td class="admin_menu"><a target="_blank" href="http://www.projecthoneypot.org/?rf=39586" title="Project Honey Pot">Project Honey Pot</a></td>
    </tr>
    </table>
EOT;

/*
 * $install_info is displayed with the title of a plugin that IS installed and
 * can be used to present extra information. In this case I show a complex
 * example that forms a button the user can click.
 */
$install_info=<<<EOT
    <table border="0" cellspacing="0" cellpadding="0"> 
    <tr>
        <td class="admin_menu"><a target="_blank" href="http://code.google.com/p/cpg-dnsbl/" title="Home">Website</a></td>
        <td class="admin_menu"><a target="_blank" href="http://tinyurl.com/36mzoa" title="Donate">Donate</a></td>
        <td class="admin_menu"><a target="_blank" href="http://www.projecthoneypot.org/?rf=39586" title="Project Honey Pot">Project Honey Pot</a></td>
    </tr>
    </table>
EOT;

?>
